specs/sharding/beacon-chain.md

```diff
@@ -730,19 +732,17 @@ def process_shard_header(state: BeaconState, signed_header: SignedShardBlobHeade
     # Charge EIP 1559 fee, builder pays for opportunity, and is responsible for later availability,
     # or fail to publish at their own expense.
     samples = body_summary.commitment.samples_count
-    # TODO: overflows, need bigger int type
-    max_fee = body_summary.max_fee_per_sample * samples
 
     # Builder must have sufficient balance, even if max_fee is not completely utilized
-    assert state.blob_builder_balances[header.builder_index] >= max_fee
+    assert state.blob_builder_balances[header.builder_index] >= body_summary.max_fee
 
-    base_fee = state.shard_sample_price * samples
+    # Round down, < 1 gwei is ok to not burn.
+    base_fee = Gwei((state.shard_sample_prices[shard] * Wei(samples)) / Wei(1000_000_000))
```

Comment:
`Wei(samples)` seems semantically incorrect

Reply:
Type checking. Alternatively we  use a bare uint256 first, and then convert to Wei at the end.

Reply:
Sorry, I would continue arguing here :) 

To be strict the operation `Wei * Wei` doesn't make any sense as it would result in a weird unit of measurement `Wei^2`. 
`Wei` type (aka additive group) should only have operations ` + Wei`, `- Wei` and scalar multiplications ` * int`, ` / int`, (resulting in `Wei` value). 
> Alternatively we use a bare uint256 first, and then convert to Wei at the end.

I believe that `Wei(123)` should be though of as short notice of `1_Wey * 123`, where `1_Wei` could be thought of as 'group generator' 

May be we can overload arithmetic operations for `Wei` (and `GWei`) in the pyspec for type safety? Kind of: 
```
def __add__(weis: Wei) Wei
def __mul__(scalar: uint64): Wei
```

Reply:
- Wei is removed, not a concern there anymore
- `A * B` already coerces the output to the type of A, or to B if A is a bare int, and checks the integer type size. But the mypy type checker has trouble with it I think, we had to change other similar cases in the pyspec-enabled specs.

Changed the code to `Gwei(uint64(state.shard_sample_prices[shard]) * samples)`, just so the mypy thing works